velox/functions/prestosql/aggregates/AverageAggregate.cpp

```diff
@@ -279,19 +298,97 @@ class AverageAggregate : public exec::Aggregate {
       } else {
         clearNull(rawNulls, i);
         auto* sumCount = accumulator(group);
-        rawValues[i] = TResult(sumCount->sum) / sumCount->count;
+        if constexpr (std::is_same_v<TAccumulator, UnscaledLongDecimal>) {
+          // Handles round-up of fraction results.
+          // TAccumulator quotient(0);
+          // DecimalUtil::divideWithRoundUp<TAccumulator, TAccumulator,
+          // int64_t>(
+          //     quotient, TAccumulator(sumCount->sum), sumCount->count, 0, 0);
+          // rawValues[i] = TResult(quotient.unscaledValue());
+          rawValues[i] = TResult(sumCount->sum.unscaledValue());
+        } else {
+          rawValues[i] = TResult(sumCount->sum) / sumCount->count;
+        }
       }
     }
   }
 
   DecodedVector decodedRaw_;
   DecodedVector decodedPartial_;
-};
+}; // class AverageAggregate
+
+template <>
+template <bool tableHasNulls>
+void AverageAggregate<
+    UnscaledShortDecimal,
+    UnscaledLongDecimal,
+    UnscaledShortDecimal>::
+    updateNonNullValue(char* group, UnscaledShortDecimal value) {
+  if constexpr (tableHasNulls) {
+    exec::Aggregate::clearNull(group);
+  }
+
+  accumulator(group)->count += 1;
+  UnscaledLongDecimal quotient(0);
+  DecimalUtil::
+      divideWithRoundUp<UnscaledLongDecimal, UnscaledLongDecimal, int64_t>(
+          quotient,
+          UnscaledLongDecimal(value) - accumulator(group)->sum,
+          accumulator(group)->count,
+          0,
+          0);
+  accumulator(group)->sum += quotient;
```

Comment:
The accumulator needs to be updated to something like this:

```c++
template <typename T>
struct Accumulator {
  T partialSum;
  int64_t partialCount;

  T totalAverage;
  double totalFraction;
  int64_t totalCount;
}
```

When adding a new value:

1. If `partialSum` is not going to overflow, add the value to `partialSum`;

2. Otherwise update `totalAverage` to integral part of `((totalAverage + totalFraction) * totalCount + partialSum) / (totalCount + partialCount)` and `totalFraction` to factional part of it.  Rearrange the formula to avoid overflow.  We call this operation `flush`.  Then we empty `partialSum` and add new value to it.

When merging 2 accumulators `x` and `y`, we `flush` both of them first, then the new `totalAverage` will be the integral part of

```
((x.totalAverage + x.totalFraction) * x.totalCount + (y.totalAverage + y.totalFraction) * y.totalCount) / (x.totalCount + y.totalCount)
```

and new `totalFraction` to be the fractional part.  Again rearrange the formula to avoid overflow.

Reply:
@Yuhta this is a bit complicated to code up. We need to break the sum into multiple parts to avoid overflow caused by multiplying `totalCount`. Also, we are trying to add totalAverage(int128_t) to totalFraction(double) here which requires casting.

If we even try to break this addition into multiple parts, we may end up loosing/rounding up fractions due to the division.

```
((totalAverage + totalFraction) * totalCount + partialSum) / (totalCount + partialCount)
```
I understand this is a good optimization to avoid division at each iteration but it is prone to deviate from the actual result.

Secondly, it doesn't solve our problem of not able to generate Partial aggregations or Intermediate aggregations.

Reply:
For the first question, the rounding in fraction part won't be a big issue as it's numerically stable.  You also mentioned an alternative way to count the number of overflows, that one can also be used if you want.
For the second question I don't quite get it, the second formula I provide is exactly for merging accumulators.  You can also derive your own if you decide to go with the overflow count method.